openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxUserAgentFactory.java

```diff
@@ -0,0 +1,95 @@
+package com.openhtmltopdf.pdfboxout;
+
+/**
+ *  PdfBoxUserAgentFactory. Singleton.
+ */
+public class PdfBoxUserAgentFactory {
+
+    private static class LazyFactoryHolder {
+        private static final PdfBoxUserAgentFactory INSTANCE = new PdfBoxUserAgentFactory();
+    }
+
+    protected static PdfBoxUserAgentFactory instance() {
+        return LazyFactoryHolder.INSTANCE;
+    }
+
+    private static class LazyHolder {
+        private static final UserAgentFactory INSTANCE = new UserAgentFactory() {
+            public PdfBoxUserAgent getPdfBoxUserAgent(PdfBoxOutputDevice outputDevice) {
+                return new PdfBoxUserAgent(outputDevice);
+            }
+        };
+    }
+
+    /*
+     * In all versions of Java, the idiom enables a safe, highly concurrent lazy initialization with good performance.
+     */
+    private static UserAgentFactory factoryDef() {
+        return LazyHolder.INSTANCE;
+    }
+
+    private ThreadLocal<UserAgentFactory> threadLocal;
```

Comment:
This smells like memory / class loader leaks in Tomcat etc. on WAR unloading... ThreadLocals are hard to do right in a servlet context.

Reply:
I have never had any problems with  ThreadLocal clases either Tomcat, Jetty, OC4J, Wildfly, Weblogic, etc. I can list / enumerate dozens of well-known / opensource projects using this standard Java class. It's available since JAVA 1.2 ;-) ThreadLocal could lead to memory leaks, but only if you don't know how to use it.